### PR TITLE
Improve DexExpectStepKind feature_test coverage

### PR DESCRIPTION
**Pull request by OCHyams**
Some of these tests are failing at the moment. This will be addressed in another
patch.

Each new test has a summary explaining its purpose.

Renamed directory expect_step_kinds -> expect_step_kind to match the command
DexExpectStepKind.

### CODE REVIEW

**Comment by OCHyams** [comment]
Yeah it's just insurance on top of `--fail-lt 1.0`. 

IMO it would be nice to have some textual output for the fail-lt mode as well as the return code... but for whatever reason at the time I didn't think we'd need it. Something like `test_name:  (score) [PASS/FAIL]`. This would make the lit tests more readable. A discussion for another time, anyway.

**Comment by TomWeaver18** [comment]
should probably add a comment about the check lines as I was very confused until I read JM's comment

unless this is common for LLVM...

File: feature_tests/commands/perfect/expect_step_kind/direction/test.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+int main()
+{
+    for (int i = 0; i < 2; ++i) {
+        i = i;
+    }
+    return 0;
+}
+
+// DexExpectStepKind('BACKWARD', 2)
+// DexExpectStepKind('FORWARD', 3)
```
Comment (TomWeaver18) [inline feature_tests/commands/perfect/expect_step_kind/direction/test.cpp]: These tests should probably have a trailing newline.

https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: feature_tests/commands/perfect/expect_step_kind/func/test.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+int main()
+{
+    func(0);
+    func(1);
+    return 0;
+}
+
+// main, func, func
+// DexExpectStepKind('FUNC', 3)
```
Comment (TomWeaver18) [inline feature_tests/commands/perfect/expect_step_kind/func/test.cpp]: missing trailing new line

File: feature_tests/commands/perfect/expect_step_kind/func_external/test.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+int main()
+{
+    func(0);
+    func(1);
+    return 0;
+}
+
+// DexExpectStepKind('FUNC_EXTERNAL', 2)
```
Comment (TomWeaver18) [inline feature_tests/commands/perfect/expect_step_kind/func_external/test.cpp]: missing trailing new line

File: feature_tests/commands/perfect/expect_step_kind/recursive/test.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+int main()
+{
+    return func(3);
+}
+
+// main, func, func, func
+// DexExpectStepKind('FUNC', 4)
+// DexExpectStepKind('BACKWARD', 0)
```
Comment (TomWeaver18) [inline feature_tests/commands/perfect/expect_step_kind/recursive/test.cpp]: missing trailing new line

File: feature_tests/commands/perfect/expect_step_kind/small_loop/test.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+int main()
+{
+    for (int i = 0; i < 2; ++i) {
+        func(i);
+    }
+    return 0;
+}
+
+// DexExpectStepKind('BACKWARD', 2)
```
Comment (TomWeaver18) [inline feature_tests/commands/perfect/expect_step_kind/small_loop/test.cpp]: missing trailing new line

**Comment by OCHyams** [comment]
@TomWeaver18 
> These tests should probably have a trailing newline.

The tests do all have a trailing newline. Github doesn't show it, but it does show a 'missing newline' icon if you're missing one.

> should probably add a comment about the check lines as I was very confused until I read JM's comment

The way I've written these matches the existing dexter feature_tests, so I think this should be address separately and for all the existing tests for consistency. In an earlier comment I said:

> IMO it would be nice to have some textual output for the fail-lt mode as well as the return code... but for whatever reason at the time I didn't think we'd need it. Something like test_name: (score) [PASS/FAIL]. This would make the lit tests more readable. A discussion for another time, anyway.

This behaviour could be added in another patch along with updated feature_tests?

**Comment by TomWeaver18** [comment]
> The tests do all have a trailing newline. Github doesn't show it, but it does show a 'missing newline' icon if you're missing one. 

ahhh that's my bad, I spotted the one with a missing new line icon and then "noticed" the others didn't have one displayed either...

how strange for the github to hide it when it's clearly been added in the test.

**Comment by TomWeaver18** [comment]
> IMO it would be nice to have some textual output for the fail-lt mode as well as the return code

agreed, we should spin off a ticket for potential future work.

**Comment by OCHyams** [comment]
Ty for the reviews.

**Comment by OCHyams** [comment]
Please see https://github.com/SNSystems/dexter/pull/59

**Pull request closed without merging**